### ISSUE TITLE
"error downloading required artifacts" is misleading when the artifact is already in the cache

### ISSUE THREAD

**Issue by cburroughs** [pex-tool/pex#2113]
Trying to debug a problem with pants `2.16.0rc0`  where an M1 mac user would get:

```
14:26:02.33 [ERROR] 1 Exception encountered:

Engine traceback:
  in `run` goal

ProcessExecutionFailure: Process 'Building 1 requirement for src.py.eq.feast_features/feast-cli.pex from the 3rdparty/py/dags.lock resolve: feast[aws,snowflake]==0.25.0' failed with exit code 1.
stdout:

stderr:
There was 1 error downloading required artifacts:
1. feast 0.25 from https://files.pythonhosted.org/packages/fc/58/7dee70ba96b0dd79d15a9378a3d327ae5930d34a21f23479391f93e898e4/feast-0.25.0.tar.gz
    hint: See above for details.
```

Lead to adding adding a bunch of `-vvvv` to the `__run.sh` script. This suggests the actual problem is:
```
Collecting https://files.pythonhosted.org/packages/fc/58/7dee70ba96b0dd79d15a9378a3d327ae5930d34a21f23479391f93e898e4/feast-0.25.0.tar.gz
  Using cached feast-0.25.0.tar.gz (3.4 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Installing backend dependencies: started
  Installing backend dependencies: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  ```

Which turns out to be a real problem with this package and M1 Mac, but that's not the direction the original error message was pointing.

Fuller log: https://gist.github.com/cburroughs/56b625630743b864a42fb0363854e389

**Comment by jsirois** [comment]
@cburroughs since Pex does provide `-v` for exactly this reason, would you be satisfied if Pants ran Pex that way, suppressing output on success, but dumping the `-v`s log on failure? Or are you wanting Pex to learn Pip / all the various ways the packages it builds can fail, parse that set of errors and produce a higher level error message?

**Comment by cburroughs** [comment]
> would you be satisfied if Pants ran Pex that way, suppressing output on success, but dumping the -vs log on failure?

I think that would have been enough to point the person (new to pants and pex) who go the error in a helpful direction, instead of trying to `curl` or looking for dns problems.

**Comment by jsirois** [comment]
Ok. So, just to make sure I'm not giving you the runaround, if I move / create a new issue over in Pants to do this then that is agreeable to you @cburroughs?

**Comment by aviau** [comment]
I just finished a long debugging session where we were debugging networking problems to realize pex wasn't downloading...

**Comment by jsirois** [comment]
@aviau, are you too a Pants user or a direct Pex user? As a direct Pex user you have easier access to debug logs (-v (up to 9 times) or `--preserve-pip-download-log`, which is where the failure to build error (which is presumably your error - you don't provide details) would show up.

**Comment by aviau** [comment]
Yeah I am a pants user, I could get the logs via `PANTS_PEX_VERBOSITY=9`, but the "error downloading required artifacts" really pointed us in the wrong direction. We had just installed a pip proxy so we were certain that this was related to our pip proxy, but it turned out to be an installation issue.

**Comment by jsirois** [comment]
Ok. As @cburroughs said here:

> But on a pendantic note in pex land, I do still think "error downloading required artifacts" as a message should be reserved for actual errors downloading

However, from Pex's perspective the error did in fact occur downloading. It just finished calling `pip download ...` on a bunch of artifacts and Pip's error message was `hint: See above for details.`. As such, the following is the current error message Pex composes:
```
1: There was 1 error downloading required artifacts:
2: 1. feast 0.25 from https://files.pythonhosted.org/packages/fc/58/7dee70ba96b0dd79d15a9378a3d327ae5930d34a21f23479391f93e898e4/feast-0.25.0.tar.gz
3:    hint: See above for details.
```

Lines 1 and 2 are from Pex, line 3 is the error message from Pip.

For Pex to do better it would have to dissect the verbose Pip logs to divine if the error was a download error (network error), or a metadata generation error (the case that caught you both up). As you might guess, this is likely non-trivial given the varietly of failure cases and the variety of Pip versions Pex supports. With two reports though I can look at at least covering the metadata generation case. Perhaps I can surface those without too-gross hacks.

**Comment by DLukeNelson** [comment]
What about something like a prefix on those lines?
Something like:
```
1: [pex] There was 1 error downloading required artifacts:
2: [pex] 1. feast 0.25 from https://files.pythonhosted.org/packages/fc/58/7dee70ba96b0dd79d15a9378a3d327ae5930d34a21f23479391f93e898e4/feast-0.25.0.tar.gz
3: [pip]   hint: See above for details.
```
would immediately have made it clear to me that I needed to get into the `pip` logs, instead of assuming that what came above (the list of requirements, and some warnings about forcibly re-creating .lck files, in my case) was the information I was told to "See above"

**Comment by jsirois** [comment]
@DLukeNelson that mostly makes sense. You're seeing things through Pants eyes though; so the `[pex] ` stuff is redundant from the Pex layer. With Pex verbose logging turnex on Pex log lines are prefixed with `pex: ` but Pex end-user output (which is output whether or not logging is enables), is of course, not. The upshot is this:
```
:; python -mpex --pip-version latest --lock lock.json -ofeast.pex
There was 1 error downloading required artifacts:
1. feast 0.25 from https://files.pythonhosted.org/packages/fc/58/7dee70ba96b0dd79d15a9378a3d327ae5930d34a21f23479391f93e898e4/feast-0.25.0.tar.gz
    pip:   error: subprocess-exited-with-error
    pip:
    pip:   × Preparing metadata (pyproject.toml) did not run successfully.
    pip:   │ exit code: 1
    pip:   ╰─> See above for output.
    pip:
    pip:   note: This error originates from a subprocess, and is likely not a problem with pip.
    pip: error: metadata-generation-failed
    pip:
    pip: × Encountered error while generating package metadata.
    pip: ╰─> See above for output.
    pip:
    pip: note: This is an issue with the package mentioned above, not pip.
    pip: hint: See above for details.
```

Still not great, since Pip buries the real error here in its own subprocess, but I think the best I can reasonably do.

**Comment by jsirois** [comment]
Ok, the better download error messages are now available in v2.13.0.
